File: RbnVfdDisplay/Models/Spot.cs
using System;

namespace RbnVfdDisplay.Models
{
    /// <summary>
    /// Represents an aggregated RBN spot for a station on a specific frequency band
    /// </summary>
    public class AggregatedSpot
    {
        /// <summary>
        /// The spotted station's callsign
        /// </summary>
        public string Callsign { get; set; } = string.Empty;

        /// <summary>
        /// Average frequency in kHz (aggregated from spots within 1 kHz of each other)
        /// </summary>
        public double FrequencyKhz { get; set; }

        /// <summary>
        /// The center frequency used to group spots (rounded to nearest kHz)
        /// </summary>
        public double CenterFrequencyKhz { get; set; }

        /// <summary>
        /// Highest SNR observed for this station/frequency combination
        /// </summary>
        public int HighestSnr { get; set; }

        /// <summary>
        /// Average speed in WPM
        /// </summary>
        public double AverageSpeed { get; set; }

        /// <summary>
        /// Total of all reported speeds (for averaging)
        /// </summary>
        public double TotalSpeed { get; set; }

        /// <summary>
        /// Total of all reported frequencies (for averaging)
        /// </summary>
        public double TotalFrequency { get; set; }

        /// <summary>
        /// Count of spots aggregated into this entry
        /// </summary>
        public int SpotCount { get; set; }

        /// <summary>
        /// Timestamp of the most recent spot for this entry
        /// </summary>
        public DateTime LastSpottedUtc { get; set; }

        /// <summary>
        /// Unique key for this aggregated spot (callsign + center frequency)
        /// </summary>
        public string Key => $"{Callsign}|{CenterFrequencyKhz:F0}";

        /// <summary>
        /// Format the spot for display on the VFD (max 24 characters)
        /// Format: "14033.0 WO6W 24"
        /// </summary>
        public string ToDisplayString()
        {
            // Frequency to 100 Hz resolution (e.g., 14033.0)
            string freqStr = $"{FrequencyKhz:F1}";
            
            // Speed as integer
            string speedStr = $"{Math.Round(AverageSpeed)}";
            
            // Calculate available space for callsign
            // Format: "FREQ.F CALL SPD" where FREQ.F is up to 8 chars, SPD is up to 3 chars
            int usedChars = freqStr.Length + 1 + speedStr.Length + 1; // freq + space + speed + space
            int callMaxLen = Math.Max(1, 24 - usedChars);
            
            string callStr = Callsign.Length <= callMaxLen 
                ? Callsign 
                : Callsign.Substring(0, callMaxLen);
            
            return $"{freqStr} {callStr} {speedStr}";
        }

        public override string ToString()
        {
            return $"{Callsign} {FrequencyKhz:F1} kHz SNR:{HighestSnr} SPD:{AverageSpeed:F0} WPM ({SpotCount} spots)";
        }
    }

    /// <summary>
    /// Represents a raw RBN spot as received from telnet
    /// </summary>
    public class RawSpot
    {
        public string SpotterCallsign { get; set; } = string.Empty;
        public string SpottedCallsign { get; set; } = string.Empty;
        public double FrequencyKhz { get; set; }
        public int Snr { get; set; }
        public int SpeedWpm { get; set; }
        public string Mode { get; set; } = string.Empty;
        public DateTime TimestampUtc { get; set; }

        public override string ToString()
        {
            return $"{SpotterCallsign} spotted {SpottedCallsign} on {FrequencyKhz:F1} kHz, SNR:{Snr} dB, {SpeedWpm} WPM";
        }
    }
}

File: rbn-vfd-linux/src/models/spot.rs
use std::time::Instant;

/// Raw spot data as received from RBN telnet
#[derive(Debug, Clone)]
pub struct RawSpot {
    pub spotter_callsign: String,
    pub spotted_callsign: String,
    pub frequency_khz: f64,
    pub snr: i32,
    pub speed_wpm: i32,
    pub mode: String,
    pub timestamp: Instant,
}

impl RawSpot {
    pub fn new(
        spotter_callsign: String,
        spotted_callsign: String,
        frequency_khz: f64,
        snr: i32,
        speed_wpm: i32,
        mode: String,
    ) -> Self {
        Self {
            spotter_callsign,
            spotted_callsign,
            frequency_khz,
            snr,
            speed_wpm,
            mode,
            timestamp: Instant::now(),
        }
    }
}

/// Aggregated spot data for display
#[derive(Debug, Clone)]
pub struct AggregatedSpot {
    pub callsign: String,
    pub frequency_khz: f64,
    pub center_frequency_khz: f64,
    pub highest_snr: i32,
    pub average_speed: f64,
    pub spot_count: u32,
    pub last_spotted: Instant,
}

impl AggregatedSpot {
    /// Create a new aggregated spot from a raw spot
    pub fn from_raw(raw: &RawSpot) -> Self {
        Self {
            callsign: raw.spotted_callsign.clone(),
            frequency_khz: raw.frequency_khz,
            center_frequency_khz: raw.frequency_khz.round(),
            highest_snr: raw.snr,
            average_speed: raw.speed_wpm as f64,
            spot_count: 1,
            last_spotted: Instant::now(),
        }
    }

    /// Update this spot with new data using incremental averaging
    pub fn update(&mut self, raw: &RawSpot) {
        self.spot_count += 1;
        self.average_speed += (raw.speed_wpm as f64 - self.average_speed) / self.spot_count as f64;
        self.frequency_khz += (raw.frequency_khz - self.frequency_khz) / self.spot_count as f64;
        if raw.snr > self.highest_snr {
            self.highest_snr = raw.snr;
        }
        self.last_spotted = Instant::now();
    }

    /// Generate the unique key for this spot (callsign + center frequency)
    pub fn key(&self) -> String {
        format!("{}|{:.0}", self.callsign, self.center_frequency_khz)
    }

    /// Format for VFD display (max 20 characters)
    /// Format: "14033.0 WO6W 24"
    pub fn to_display_string(&self) -> String {
        let freq_str = format!("{:.1}", self.frequency_khz);
        let speed_str = format!("{}", self.average_speed.round() as i32);

        // Calculate available space for callsign
        // Format: "FREQ.F CALL SPD"
        let used_chars = freq_str.len() + 1 + speed_str.len() + 1;
        let call_max_len = 20_usize.saturating_sub(used_chars);

        let call_str = if self.callsign.len() <= call_max_len {
            &self.callsign
        } else {
            &self.callsign[..call_max_len]
        };

        format!("{} {} {}", freq_str, call_str, speed_str)
    }
}
